projects/table-builder/src/lib/classes/table-store.ts

```diff
@@ -0,0 +1,192 @@
+import { FieldType, MetaData } from '../interfaces/report-def';
+import { v4 as uuid } from 'uuid';
+import { Observable } from 'rxjs';
+import { defaultTableState, TableState } from './TableState';
+import { Injectable, Inject } from '@angular/core';
+import { TableBuilderConfig, TableBuilderConfigToken } from './TableBuilderConfig';
+import { FilterInfo } from './filter-info';
+import { selectStorageStateItem, StateStorage } from '../ngrx/reducer';
+import { Sort, SortDirection }  from '@angular/material/sort' ;
+import { ComponentStore } from '@ngrx/component-store'  ;
+import update from 'immutability-helper';
+import { Dictionary } from '../interfaces/dictionary';
+import { select, Store } from '@ngrx/store';
+import { first, mergeMap, tap } from 'rxjs/operators';
+import { loadState, saveState } from '../ngrx/actions';
+
+
+@Injectable()
+export class TableStore extends ComponentStore<TableState> {
+
+  constructor(@Inject(TableBuilderConfigToken) private config: TableBuilderConfig,
+  private store: Store<{storageState: StateStorage}>) {
+    super( { ...defaultTableState, ...config.defaultTableState});
+  }
+  readonly filters$ = this.select(state => state.filters );
+
+  getFilter$(filterId) : Observable<FilterInfo | undefined> {
+    return this.select( this.filters$, filters => filters[filterId]);
+  }
+
+
+  setFromSavedState(id:string) {
+    this.store.dispatch(loadState({id}));
+    this.updateState( this.store.pipe(
+      select(selectStorageStateItem(id)),
+      mergeMap( (state: TableState ) => (state ? [state] : [] )),
+      first(),
+    ));
+  }
+
+  on<T>( srcObservable: Observable<T>, func: (obj:T)=> void) {
+    this.effect((src: Observable<T>) => {
+      return src.pipe(tap(func));
+    })(srcObservable);
+  }
+
+  async saveToState(id:string) {
+    const state = await this.state$.pipe(first()).toPromise();
```

Comment:
not sure I follow, but why `toPromise()`?

Reply:
the alternative would be to subscribe with a callback.
This way is more straight forward.